pkg/kp/rcstore/consul_store.go

```diff
@@ -26,39 +24,66 @@ type consulKV interface {
 	Get(key string, opts *api.QueryOptions) (*api.KVPair, *api.QueryMeta, error)
 	Keys(prefix string, separator string, opts *api.QueryOptions) ([]string, *api.QueryMeta, error)
 	List(prefix string, opts *api.QueryOptions) (api.KVPairs, *api.QueryMeta, error)
-	Put(pair *api.KVPair, opts *api.WriteOptions) (*api.WriteMeta, error)
+	CAS(pair *api.KVPair, opts *api.WriteOptions) (bool, *api.WriteMeta, error)
+	DeleteCAS(pair *api.KVPair, opts *api.WriteOptions) (bool, *api.WriteMeta, error)
 	Acquire(pair *api.KVPair, opts *api.WriteOptions) (bool, *api.WriteMeta, error)
-	DeleteTree(prefix string, opts *api.WriteOptions) (*api.WriteMeta, error)
 }
 
 type consulStore struct {
 	applicator labels.Applicator
 	kv         consulKV
 	logger     logging.Logger
+	retries    int
+}
+
+// TODO: combine with similar CASError type in pkg/labels
+type CASError string
+
+func (e CASError) Error() string {
+	return fmt.Sprintf("Could not compare-and-set key %q", string(e))
 }
 
 var _ Store = &consulStore{}
 
-func NewConsul(client *api.Client) *consulStore {
+func NewConsul(client *api.Client, retries int, logger logging.Logger) *consulStore {
 	return &consulStore{
-		// TODO: Should the number of retries be configurable?
-		applicator: labels.NewConsulApplicator(client, 3),
+		retries:    retries,
+		applicator: labels.NewConsulApplicator(client, retries),
 		kv:         client.KV(),
-		logger:     logging.DefaultLogger,
+		logger:     logger,
 	}
 }
 
 func (s *consulStore) Create(manifest pods.Manifest, nodeSelector labels.Selector, podLabels labels.Set) (fields.RC, error) {
-	id := fields.ID(uuid.New())
+	rc, err := s.innerCreate(manifest, nodeSelector, podLabels)
+	for i := 0; i < s.retries; i++ {
+		if _, ok := err.(CASError); ok {
+			rc, err = s.innerCreate(manifest, nodeSelector, podLabels)
+		} else {
+			break
+		}
+	}
+	if err != nil {
+		return fields.RC{}, err
+	}
 
-	buf := bytes.Buffer{}
-	err := manifest.Write(&buf)
+	// labels do not need to be retried, consul applicator does that itself
+	err = s.forEachLabel(rc, func(id, k, v string) error {
```

Comment:
I don't feel strongly but maybe consider retrying here instead of within the applicator? Depends on how many things are going to use the applicator. Probably good the way it is